horovod/common/operations.cc

```diff
@@ -1010,7 +1001,20 @@ Status EnqueueTensorAllreduces(std::vector<std::shared_ptr<OpContext>>& contexts
     e.callback = std::move(callbacks[n]);
 
     entries.push_back(std::move(e));
+  }
 
+  // Start appropriate NVTX range
+  if (tensors.size() == 1) {
+    auto& e = entries[0];
+    e.nvtx_op_range.Start(RegisteredNvtxOp::HorovodAllreduce, e.tensor->size());
+  } else {
+    auto total_size =
+        std::accumulate(entries.begin(), entries.end(), 0ll,
+                        [](int64_t size_sum, const TensorTableEntry& e) {
+                          return size_sum + e.tensor->size();
+                        });
+    auto& e = entries[tensors.size() - 1];
```

Comment:
There is a chance that this range for the grouped allreduce might be slightly inaccurate in some cases. There is a guarantee that the tensors submitted in a group will be involved in collectives executed on the same cycle, but there is no guarantee that they will all be a part of the same call to `PerformOperation`. For example, if the group contains a set of entries with a combined size that exceeds `HOROVOD_FUSION_THRESHOLD`, that will result in multiple calls to `PerformOperation` on the separate fused buffers. Subsequently, the NVTX range will be ended by whichever of those `PerformOperation` calls contains `entries[tensor.size() - 1]`, which isn't necessarily the last one executed. 

Perhaps there is a way to get around this by having the entries of the group share ownership of a range (maybe store a `shared_ptr` to the range in the tensor table entry), and have the range end on object destruction?

Reply:
Thanks for catching that mistake! 

I ended up with an opaque wrapper around a `shared_ptr` to a range object. This way the compiler can optimize it away if we build with `HOROVOD_WITHOUT_NVTX=1`.